Compatibility with Doctrine ORM 3.0 and DBAL 4.0

ondrejmirtes opened the pull request.

ondrejmirtes committed:
Build with Doctrine ORM 3 and DBAL 4 in CI

ondrejmirtes committed:
Add AnnotationDriver so that ORM PHPDoc annotations can be used with Doctrine ORM 3.0

ondrejmirtes committed:
Support Composer patches

ondrejmirtes committed:
Patch Attribute classes to support AnnotationDriver even with Doctrine ORM 3 for unit tests

VincentLanglet commented on tests/bootstrap.php:
```diff
@@ -5,3 +5,10 @@
+	!is_file(__DIR__ . '/../vendor/doctrine/orm/src/Mapping/Driver/AnnotationDriver.php')
+	&& !is_file(__DIR__ . '/../vendor/doctrine/orm/lib/Doctrine/ORM/Mapping/Driver/AnnotationDriver.php')
+) {
+	require_once __DIR__ . '/../compatibility/AnnotationDriver.php';
```
This is a nice idea indeed in order to have less code to update currently.

But I feel like it could still be useful in long term (certainly in another PR which can be merged before or after the doctrine bump) to add Attributes everywhere Annotation is used in tests in order to tests them too. Also, rather having some time `data/` and `data-attributes` folders with duplicated tests, a unique `data/` folder can be used with both annotation and attribute in the same class without any issue.

This way the logic
```
if (! class_exists(AnnotationDriver::class) {
    // use AttributeReader
} else {
   // use AnnotationReader (or decide based on PHP version if wanted)
}
```
can be use in the entity-manager.php.

ondrejmirtes replied:
> to add Attributes everywhere Annotation

I disagree with this approach. I consider existing tests to be immutable. By changing them we might start testing different things by accident than originally intended.

ondrejmirtes committed:
Fix PropertiesExtension

ondrejmirtes committed:
ORMException moved to another namespace

VincentLanglet commented:
For all the 
```
Method PHPStan\Type\Doctrine\Query\QueryResultTypeWalker::walkEntityIdentificationVariable() has parameter $identVariable with no type specified.
```
PHPStan errors, it's because the `inheritdoc` phpdoc is useless ; doctrine/orm 3 use typehint and no phpdoc.

I fixed them by adding `@param` on every QueryResultTypeWalker method.

I can extract the commit from https://github.com/phpstan/phpstan-doctrine/pull/531 if you want to cherry pick it.

(also, do you want / can I help for something ?)

ondrejmirtes commented:
Please do, that would be helpful 😊

VincentLanglet committed:
Remove inheritdoc phpdoc

VincentLanglet commented:
> Please do, that would be helpful 😊

Commit is here https://github.com/phpstan/phpstan-doctrine/commit/3ebf91ff0f34938c50157480fd33aeee76f78fee

ondrejmirtes commented:
Can you send it as a PR with base branch doctrine-orm-3, so I can just merge it from my phone? Thanks.

VincentLanglet commented:
> Can you send it as a PR with base branch doctrine-orm-3, so I can just merge it from my phone? Thanks.

https://github.com/phpstan/phpstan-doctrine/pull/538

ondrejmirtes committed:
More QueryResultTypeWalker fixes

ondrejmirtes commented:
Signing off today, if you could look at any of these failures https://github.com/phpstan/phpstan-doctrine/actions/runs/7833385749 @VincentLanglet, I'd appreciate it :) Thanks.

VincentLanglet committed:
More fixes

VincentLanglet commented:
> Signing off today, if you could look at any of these failures [phpstan/phpstan-doctrine/actions/runs/7833385749](https://github.com/phpstan/phpstan-doctrine/actions/runs/7833385749) @VincentLanglet, I'd appreciate it :) Thanks.

I fixed some tests on https://github.com/phpstan/phpstan-doctrine/pull/540

I let you the PHPStan error, since I'm not sure how you prefer to fix them.

lugosium commented:
I tried this branch and everythings work fine for me

`645/645 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100% 30 secs`

👏

ondrejmirtes committed:
Fix test

ondrejmirtes committed:
Tests - bring back DBAL ArrayType

ondrejmirtes committed:
Patch Carbon return types

ondrejmirtes committed:
Fix tests for DBAL 4

ondrejmirtes committed:
Fix tests for ORM 3

ondrejmirtes committed:
Skip tests not working on ORM 3

Related: https://github.com/doctrine/orm/issues/11240

ondrejmirtes committed:
Fix more tests on ORM 3

ondrejmirtes committed:
BigIntType - always uses integer in DBAL 4

ondrejmirtes committed:
EntityColumnRuleTest - testGeneratedIds with less complicated type than bigint

ondrejmirtes committed:
More patches for Doctrine bugs

Related: https://github.com/doctrine/orm/issues/11240
Related: https://github.com/doctrine/orm/issues/11241

ondrejmirtes committed:
Fix PHPStan

arderyp commented:
hey @ondrejmirtes it looks like this was completed and merged into `1.3.61`.

I'm running:

phpstan-doctrine `1.3.62`
orm `3.1.0`
dbal `3.8.3`

When I upgraded `orm` from latest 2.x to `3.0.x` and `3.1.x`, I got lots of errors about `EntityManager` methods returning mixed.

Am I missing something, or is orm3 only supported WITH dbal4?

I opened [another issue](https://github.com/phpstan/phpstan-doctrine/issues/551) before finding this one.

EDIT: bumping to dbal4 is throwing different errors now:

```
1/1 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 -- -------------------------------------------------------------------------------------------------------------------------------------------- 
     Error                                                                                                                                       
 -- -------------------------------------------------------------------------------------------------------------------------------------------- 
     Internal error: Internal error: Doctrine\DBAL\Connection\StaticServerVersionProvider::__construct(): Argument #1 ($version) must be of      
     type string, float given, called in /app/vendor/doctrine/dbal/src/Connection.php on line 186 while  
     analysing file /app/src/Common/Repository/ApplicationsRepository.php                                
     Run PHPStan with -v option and post the stack trace to:                                                                                     
     https://github.com/phpstan/phpstan/issues/new?template=Bug_report.yaml                                                                      
     Child process error (exit code 1):                                                                                                          
 -- -------------------------------------------------------------------------------------------------------------------------------------------- 
                                                                                                               
 [ERROR] Found 2 errors
```
